File: python/taflow/klinger_volume_oscillator.py
"""Native Klinger volume oscillator interface."""

from typing import Any

import numpy as np

from ._native import StatefulKlingerVolumeOscillator


class KlingerVolumeOscillator:
    """Compute signed-volume force with fast/slow EMA and signal EMA.

    Parameters
    ----------
    high, low, close, volume : array-like
        Initial aligned OHLCV history.
    fast, slow, signal : int, default 34, 55, 13
        EMA periods for force, baseline, and signal smoothing.
    """

    def __init__(
        self,
        high: Any,
        low: Any,
        close: Any,
        volume: Any,
        fast: int = 34,
        slow: int = 55,
        signal: int = 13,
    ) -> None:
        """Initialize this adapter and process the supplied input series.

        Parameters
        ----------
        high : object
            High-price series or the current bar high.
        low : object
            Low-price series or the current bar low.
        close : object
            Close-price series or the current bar close.
        volume : object
            Volume series or the current bar volume.
        fast : object
            Fast smoothing period in bars.
        slow : object
            Slow smoothing period in bars.
        signal : object
            Signal smoothing period in bars.

        Returns
        -------
        None
            The constructor initializes the adapter and returns no value.
        """
        self._state = StatefulKlingerVolumeOscillator(fast, slow, signal)
        if close is not None:
            self.extend(high, low, close, volume)

    def append(self, high: float, low: float, close: float, volume: float) -> "KlingerVolumeOscillator":
        """Append one chronological observation to the native Rust state.

        Parameters
        ----------
        high : float
            Current high price.
        low : float
            Current low price.
        close : float
            Current close price.
        volume : float
            Current volume.

        Returns
        -------
        KlingerVolumeOscillator
            This indicator, for fluent chaining; read `value` for the result."""
        self._state.append(float(high), float(low), float(close), float(volume))
        return self

    def extend(self, high: Any, low: Any, close: Any, volume: Any) -> "KlingerVolumeOscillator":
        """Append aligned chronological histories to the native Rust state.

        Parameters
        ----------
        high : Any
            Chronological high price series.
        low : Any
            Chronological low price series.
        close : Any
            Chronological close price series.
        volume : Any
            Chronological volume series.

        Returns
        -------
        KlingerVolumeOscillator
            This indicator, for fluent chaining."""
        self._state.extend(
            np.asarray(high, dtype=np.float64),
            np.asarray(low, dtype=np.float64),
            np.asarray(close, dtype=np.float64),
            np.asarray(volume, dtype=np.float64),
        )
        return self

    def compute(self) -> object:
        """Return the complete aligned history produced by Rust.

        Returns
        -------
        numpy.ndarray or tuple of numpy.ndarray
            One output per processed bar, including NaN warm-up positions."""
        return self._state.compute()

    @property
    def value(self) -> object:
        """Return the latest Rust result.

        Returns
        -------
        float, tuple, or None
            Latest output, or None while scalar warm-up is incomplete."""
        return self._state.value

    def reset(self) -> "KlingerVolumeOscillator":
        """Restore fresh-state behavior and clear output history.

        Returns
        -------
        KlingerVolumeOscillator
            This indicator, for fluent chaining."""
        self._state.reset()
        return self

File: crates/taflow-core/src/stream/klinger_volume_oscillator.rs
//! Stateful Klinger volume oscillator.

use super::invalid_period;
use crate::error::TaResult;

/// Fast/slow EMA difference of signed volume force and its signal EMA.
#[derive(Debug, Clone)]
/// Persistent Rust state or aligned output type for `KlingerVolumeOscillator`.
///
/// The state consumes chronological inputs causally, preserves warm-up
/// values, and exposes the current result through its public API.
pub struct KlingerVolumeOscillator {
    previous_typical_price: Option<f64>,
    fast_average: SeededEma,
    slow_average: SeededEma,
    signal_average: SeededEma,
    value: Option<(f64, f64)>,
}

#[derive(Debug, Clone)]
struct SeededEma {
    period: usize,
    count: usize,
    sum: f64,
    value: Option<f64>,
}

impl SeededEma {
    fn new(period: usize) -> Self {
        Self {
            period,
            count: 0,
            sum: 0.0,
            value: None,
        }
    }

    fn append(&mut self, input: f64) -> Option<f64> {
        self.count += 1;
        self.value = if self.count < self.period {
            self.sum += input;
            None
        } else if self.count == self.period {
            self.sum += input;
            Some(self.sum / self.period as f64)
        } else {
            let alpha = 2.0 / (self.period as f64 + 1.0);
            self.value
                .map(|previous| previous + alpha * (input - previous))
        };
        self.value
    }

    fn reset(&mut self) {
        self.count = 0;
        self.sum = 0.0;
        self.value = None;
    }
}

impl KlingerVolumeOscillator {
    /// Creates the oscillator with positive fast, slow, and signal periods.
    pub fn new(fast: usize, slow: usize, signal: usize) -> TaResult<Self> {
        for (name, period) in [("fast", fast), ("slow", slow), ("signal", signal)] {
            if period < 1 {
                return Err(invalid_period(name, period, 1));
            }
        }
        Ok(Self {
            previous_typical_price: None,
            fast_average: SeededEma::new(fast),
            slow_average: SeededEma::new(slow),
            signal_average: SeededEma::new(signal),
            value: None,
        })
    }

    /// Appends one OHLCV bar and returns oscillator and signal values.
    pub fn append(&mut self, high: f64, low: f64, close: f64, volume: f64) -> (f64, f64) {
        let typical_price = (high + low + close) / 3.0;
        let trend = self.previous_typical_price.map_or(1.0, |previous| {
            if typical_price > previous {
                1.0
            } else if typical_price < previous {
                -1.0
            } else {
                0.0
            }
        });
        let force = trend * volume;
        self.previous_typical_price = Some(typical_price);
        let fast_average = self.fast_average.append(force);
        let slow_average = self.slow_average.append(force);
        let oscillator = fast_average
            .zip(slow_average)
            .map_or(f64::NAN, |(fast, slow)| fast - slow);
        let signal_average = if oscillator.is_nan() {
            f64::NAN
        } else {
            self.signal_average.append(oscillator).unwrap_or(f64::NAN)
        };
        let result = (oscillator, signal_average);
        self.value = Some(result);
        result
    }

    /// Returns the latest oscillator and signal pair.
    pub fn value(&self) -> Option<(f64, f64)> {
        self.value
    }
    /// Clears all EMA and previous-price state.
    pub fn reset(&mut self) {
        self.previous_typical_price = None;
        self.fast_average.reset();
        self.slow_average.reset();
        self.signal_average.reset();
        self.value = None;
    }
}
